test/metabase/api/dataset_test.clj

```diff
@@ -107,6 +108,44 @@
                   :running_time true}
                  (format-response (most-recent-query-execution-for-query query)))))))))
 
+(deftest override-metadata-query-test
+  (testing "POST /api/dataset"
+    (testing "when `:dataset-metadata` is passed in the body, make sure the returned columns metadata get combined with it (#22517)"
+      (let [dataset-metadata [{:name          "ID"
+                               :display_name  "VENUES ID CAPITALIZED"
+                               :description   "normal venues id but capitalized"
+                               :semantic_type "type/FK" ;; originally is type/PK
+                               :base_type     "type/BigInteger"
+                               :field_ref     ["field" (mt/id :venues :id) nil]}]
+            query            (assoc (mt/mbql-query venues {:fields [$id]})
+                                    :dataset-metadata dataset-metadata)]
+        (is (= dataset-metadata
+               (->> (mt/user-http-request :rasta :post 202 "dataset" query)
+                    mt/cols
+                    (map #(select-keys % [:name :display_name :description :base_type :field_ref :semantic_type])))))
+
+        (testing "We shouldn't keep the dataset-metadata inside query when running it"
+          (let [run-query-async api.dataset/run-query-async]
+            (with-redefs [api.dataset/run-query-async (fn [query & opts]
+                                                        (assert (nil? (:dataset-metadata query)))
+                                                        (run-query-async query opts))]
+              (mt/user-http-request :rasta :post 202 "dataset" query))))))
```

Comment:
This is also not testing anything other than that the query runs... it could easily be running just fine for some other reason and completely ignore the metadata you're passing

Reply:
keep in mind because the QP endpoint is async it always returns 202 whether it succeeds or not... you need to look at the response